peakina/cache.py

```diff
@@ -0,0 +1,131 @@
+from abc import ABCMeta, abstractmethod
+from contextlib import suppress
+from datetime import timedelta
+from enum import Enum
+from pathlib import Path
+from time import time
+
+import pandas as pd
+
+
+class CacheEnum(str, Enum):
+    MEMORY = 'memory'
+    HDF = 'hdf'
+
+
+class Cache(metaclass=ABCMeta):
+    @staticmethod
+    def get_cache(kind: CacheEnum, *args, **kwargs):
+        return {CacheEnum.MEMORY: InMemoryCache, CacheEnum.HDF: HDFCache}[kind](*args, **kwargs)
+
+    @staticmethod
+    def should_invalidate(
+        *,
+        mtime: float = None,
+        cached_mtime: float,
+        expire: timedelta = None,
+        cached_created_at: float,
+    ):
+        now = time()
+        is_newer_version = False
+        is_expired = False
+        if mtime is not None:
+            is_newer_version = mtime > cached_mtime
+        if expire is not None:
+            is_expired = now > cached_created_at + expire.total_seconds()
+        return is_newer_version or is_expired
+
+    @abstractmethod
+    def get(
+        self, key: str, mtime=None, expire: timedelta = None
+    ) -> pd.DataFrame:  # pragma: no cover
+        pass
+
+    @abstractmethod
+    def set(self, key: str, value: pd.DataFrame, mtime=None):  # pragma: no cover
+        pass
+
+    @abstractmethod
+    def delete(self, key: str):  # pragma: no cover
+        pass
+
+
+class InMemoryCache(Cache):
+    def __init__(self):
+        self._cache = {}
+
+    def get(self, key: str, mtime=None, expire: timedelta = None) -> pd.DataFrame:
+        cached = self._cache[key]
+        if self.should_invalidate(
+            mtime=mtime,
+            cached_mtime=cached['mtime'],
+            expire=expire,
+            cached_created_at=cached['created_at'],
+        ):
+            self.delete(key)
+        return self._cache[key]['value']
+
+    def set(self, key: str, value: pd.DataFrame, mtime=None):
+        mtime = mtime or time()
+        self._cache[key] = {'value': value, 'mtime': mtime, 'created_at': time()}
+
+    def delete(self, key: str):
+        if key in self._cache:
+            del self._cache[key]
+
+
+class HDFCache(Cache):
+    META_DF_KEY = '__meta__'
+
+    def __init__(self, cache_dir: Path):
```

Comment:
I think it should be a `str` here or `Union[str, Path]` and `self.cache_dir = Path(cache_dir)` !
In your tests you use the `tmp_path` fixture, which is a `pathlib.Path` instance but I think most people will want to write `my_cache = Cache.get_cache('hdf', cache_dir='/data/tmp...')` for example